app/src/main/java/com/dkanada/gramophone/service/QueueManager.java

```diff
@@ -0,0 +1,314 @@
+    }
+
+    public List<Song> getPlayingQueue() {
+        return shuffleMode == SHUFFLE_MODE_SHUFFLE ? shuffledQueue : playingQueue;
+    }
+
+    public int getPosition() {
+        return position;
+    }
+
+    public Song getCurrentSong() {
+        return getSongAt(getPosition());
+    }
+
+    public Song getSongAt(int position) {
+        if (position >= 0 && position < getPlayingQueue().size()) {
+            return getPlayingQueue().get(position);
+        }
+
+        return null;
+    }
+
+    public void setPosition(int position) {
+        this.position = position;
+    }
+
+    public void setNextPosition() {
+        // manual switch to next song
+        switch (getRepeatMode()) {
+            case REPEAT_MODE_NONE:
+            case REPEAT_MODE_THIS:
+                position = Math.min(position + 1, playingQueue.size() - 1);
+                break;
+            case REPEAT_MODE_ALL:
+                position = (position + 1) % playingQueue.size();
+                break;
+        }
+    }
+
+    public void setPreviousPosition() {
+        // manual switch to previous song
+        switch (getRepeatMode()) {
+            case REPEAT_MODE_NONE:
+            case REPEAT_MODE_THIS:
+                position = Math.max(position - 1, 0);
+                break;
+            case REPEAT_MODE_ALL:
+                position = (position - 1 + playingQueue.size()) % playingQueue.size();
+                break;
+        }
+    }
+
+    public int getRepeatMode() {
+        return repeatMode;
+    }
+
+    public int getShuffleMode() {
+        return shuffleMode;
+    }
+
+    public void toggleShuffle() {
+        if (getShuffleMode() == SHUFFLE_MODE_NONE) {
+            setShuffleMode(SHUFFLE_MODE_SHUFFLE);
+        } else {
+            setShuffleMode(SHUFFLE_MODE_NONE);
+        }
+    }
+
+    public void setRepeatMode(final int repeatMode) {
+        switch (repeatMode) {
+            case REPEAT_MODE_NONE:
+            case REPEAT_MODE_ALL:
+            case REPEAT_MODE_THIS:
+                this.repeatMode = repeatMode;
+                PreferenceUtil.getInstance(context).setRepeat(repeatMode);
+                callbacks.onRepeatModeChanged();
+                break;
+        }
+    }
+
+    public void setShuffleMode(final int shuffleMode) {
+        PreferenceUtil.getInstance(context).setShuffle(shuffleMode);
+
+        switch (shuffleMode) {
+            case SHUFFLE_MODE_SHUFFLE:
+                this.shuffleMode = shuffleMode;
+                shuffleQueue();
+
+                break;
+            case SHUFFLE_MODE_NONE:
+                String currentSongId = getCurrentSong().id;
+                int newPosition = 0;
+
+                // FixMe: what if same song more than once?
+                for(Song song: playingQueue) {
+                    if (song.id.equals(currentSongId)) {
+                        newPosition = playingQueue.indexOf(song);
+                        break;
+                    }
+                }
```

Comment:
Since we're moving this, might be cleaner to use a stream here.

Reply:
Not sure if it's cleaner, but I did use streams for now.